### PR TITLE
Add amplitude and angular quantum embeddings

### LINKED ISSUE #2982: Add common quantum embeddings
### Required prerequisites

- [x] Search the [issue tracker](https://github.com/NVIDIA/cuda-quantum/issues) to check if your feature has already been mentioned or rejected in other issues.

### Describe the feature

Add built-in APIs for common quantum embeddings.

--

**(1) Amplitude encoding:**
Maps and normalizes classical features (`list`, `ndarray` or `cudaq.State`) to a quantum state by encoding values as amplitudes:  $x \mapsto \sum_i x_i \|i\rangle$, including `pad` as an argument for vectors not sized to $2^n$ with the specified pad value, e.g., zero-padding to the nearest power of $2^n$ by setting `pad = 0`. The featured API looks like:
```
state_vector = cudaq.amplitude_encode([0.5, 0.5, 0.5], pad = 0)
```
Returns `cudaq.State` with values `[0.5773, 0.5773, 0.5773, 0]` (L2-norm = 1)

--

**(2) Angular encoding:**
Encodes classical features (`list` or `ndarray`) as qubit rotations (e.g., with specified Ry gates):  $x_i \mapsto R_y(x_i)\|0\rangle$:
```
@cudaq.kernel
def kernel():
    q = cudaq.amplitude_encode([0.1, 0.2, 0.3], rotation='Y')
```
The corresponding kernel:
```
print(cudaq.draw(kernel))
     ╭─────────╮
q0 : ┤ ry(0.1) ├
     ├─────────┤
q1 : ┤ ry(0.2) ├
     ├─────────┤
q2 : ┤ ry(0.3) ├
     ╰─────────╯
``` 

--

Both embedding methods are commonly used in QML, enabling developers to design models more easily and intuitively.

### PR DESCRIPTION
**Pull request by ssmswapnil**
## Add common quantum embeddings

Closes #2982

### Summary

Adds built-in APIs for two common quantum embedding methods used in Quantum Machine Learning (QML):

1. **`cudaq.amplitude_encode(data, pad=None)`** — Encodes classical feature vectors as quantum state amplitudes, with automatic L2-normalization and optional padding to the nearest power of 2.

2. **`cudaq.angular_encode(kernel, qubits, data, num_qubits, rotation='Y')`** — Encodes classical features as single-qubit rotation angles (Rx, Ry, or Rz), following the same builder-API pattern as `cudaq.kernels.hwe`.

### Design Decisions

- **Separated into two distinct functions** rather than a single overloaded function, since amplitude encoding and angular encoding have fundamentally different signatures and return types (`amplitude_encode` returns a `cudaq.State`, while `angular_encode` modifies a kernel in-place).
- **`angular_encode` follows the `hwe` convention** — takes `kernel` and `qubits` arguments and appends gates via the builder API, consistent with existing kernel library functions.
- **`angular_encode` supports parameterized kernels** — works with both concrete `list[float]` data and `QuakeValue` parameters for variational workflows.
- **Top-level access uses lazy loading** — `amplitude_encode` and `angular_encode` are registered in `_LAZY_ATTRS`, following the existing PEP 562 pattern used by `evolve`, `Schedule`, etc.

### Examples

```python
import cudaq

# (1) Amplitude encoding
state = cudaq.amplitude_encode([0.5, 0.5, 0.5], pad=0)
# Returns cudaq.State with amplitudes [0.5773, 0.5773, 0.5773, 0.0]

# (2) Angular encoding
kernel = cudaq.make_kernel()
qubits = kernel.qalloc(3)
cudaq.kernels.angular_encode(kernel, qubits, [0.1, 0.2, 0.3], 3, rotation='Y')
print(cudaq.draw(kernel))
#      ╭─────────╮
# q0 : ┤ ry(0.1) ├
#      ├─────────┤
# q1 : ┤ ry(0.2) ├
#      ├─────────┤
# q2 : ┤ ry(0.3) ├
#      ╰─────────╯
```

### Changes

| File | Action | Description |
|------|--------|-------------|
| `python/cudaq/kernels/embeddings.py` | New | Implementation of `amplitude_encode` and `angular_encode` |
| `python/tests/test_embeddings.py` | New | 20 unit tests covering both functions |
| `python/cudaq/kernels/__init__.py` | Modified | Added import for new embeddings module |
| `python/cudaq/__init__.py` | Modified | Added `_LAZY_ATTRS` entries for top-level access |

### Testing

All 20 tests pass:
- 11 tests for `amplitude_encode` (normalization, padding, numpy input, complex values, edge cases, error handling)
- 9 tests for `angular_encode` (Rx/Ry/Rz axes, case insensitivity, parameterized kernels, error handling)

### CODE REVIEW

**Comment by 1tnguyen** [comment]
Hi @ssmswapnil,

Thank you for your contribution! 

There are a couple of things to fix for the CI: 

- We need all commits to be signed off. Please take a look at the [instructions](https://github.com/NVIDIA/cuda-quantum/pull/4666/checks?check_run_id=79455938590) to rectify that. 

- The code format, spelling, and license header checks are failing. You can follow the instructions [here](https://github.com/NVIDIA/cuda-quantum/blob/e1eca9b34c465206a5d80788f02cf8fe919b2e93/Developing.md?plain=1#L81) to check and fix them locally.

**Comment by 1tnguyen** [comment]
> /ok to test b81d1014fd4a8837c616bc2aef045b19fb5a2048

**Command Bot:** Processing...

**Comment by ssmswapnil** [comment]
Hi @1tnguyen, thanks for the review!

I've pushed both the fixes:
- Added commit sign-off
- Reformatted files
- 
Let me know if anything else needs to be addressed.

**Comment by schweitzpgi** [comment]
C++?

**Comment by ssmswapnil** [comment]
Hi @schweitzpgi ,
The issue examples were Python-only, and I followed the existing pattern from hwe.py and uccsd.py which are also Python-only. I'm not familiar enough with the C++ side of the codebase to add that confidently. Would a Python-only implementation work for this issue, or is C++ needed?

**Comment by bettinaheim** [comment]
@ssmswapnil Thank you for this contribution!
- Can you please update the PR to have this go into the contrib module/contrib namespace?
- We'll help to unify this contribution with #4667, which contains C++ support and lists the API in the docs. 
We appreciate if you are willing to help bring this across the finish line in a collaborative manner.

**Comment by ssmswapnil** [comment]
Hi @bettinaheim ,
Thanks for the feedback! I've moved the implementation to the contrib module. I've also looked at #4667 and am happy to help unify the two contributions. Let me know how you'd like to coordinate.

**Comment by schweitzpgi** [comment]
> Hi @schweitzpgi , The issue examples were Python-only, and I followed the existing pattern from hwe.py and uccsd.py which are also Python-only. I'm not familiar enough with the C++ side of the codebase to add that confidently. Would a Python-only implementation work for this issue, or is C++ needed?

Ideally, we want parity. Understood that it isn't always possible as a first step.

**Comment by schweitzpgi** [comment]
> /ok to test 45998fa

**Command Bot:** Processing...

File: python/cudaq/contrib/embeddings.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
+# ============================================================================ #
+# Copyright (c) 2022 - 2026 NVIDIA Corporation & Affiliates.                   #
```
Comment (schweitzpgi) [inline python/cudaq/contrib/embeddings.py]:
```suggestion
# Copyright (c) 2026 NVIDIA Corporation & Affiliates.                          #
```
nit

**Comment by ssmswapnil** [comment]
Hi @schweitzpgi, thanks for the review and the suggestion. I applied the copyright header change, which created a new commit on the PR branch. It looks like the workflows for the latest commit are currently waiting for approval. Could you please approve them when you get a chance? Thanks!

**Comment by schweitzpgi** [comment]
> /ok to test cc1eec5

**Command Bot:** Processing...

**Comment by ssmswapnil** [comment]
Hi @schweitzpgi , could u please approve the workflows.
